rest-client/src/test/java/org/sagebionetworks/bridge/rest/UserSessionInfoProviderTest.java

```diff
@@ -0,0 +1,157 @@
+import org.sagebionetworks.bridge.rest.model.SignIn;
+import org.sagebionetworks.bridge.rest.model.UserSessionInfo;
+
+import retrofit2.Call;
+import retrofit2.Response;
+
+@RunWith(PowerMockRunner.class)
+@PrepareForTest({ AuthenticationApi.class, Call.class, Response.class })
+public class UserSessionInfoProviderTest {
+
+    @Mock
+    AuthenticationApi authenticationApi;
+    
+    @Mock
+    Call<UserSessionInfo> call;
+    
+    @Mock
+    Call<UserSessionInfo> call2;
+    
+    @Mock
+    Response<UserSessionInfo> response;
+    
+    @Captor
+    ArgumentCaptor<ReauthenticateRequest> reauthenticateRequestCaptor;
+
+    @Captor
+    ArgumentCaptor<SignIn> signInCaptor;
+    
+    SignIn signIn;
+    
+    UserSessionInfo userSessionInfo;
+    
+    UserSessionInfoProvider provider;
+    
+    @Before
+    public void before() {
+        userSessionInfo = new UserSessionInfo()
+                .email("email@email.com")
+                .sessionToken("sessionToken");
+        
+        signIn = new SignIn().email("email@email.com").password("password").study("test-study");
+        
+        provider = new UserSessionInfoProvider(authenticationApi, signIn);
+    }
+    
+    @Test
+    public void onceItHasSessionItDoesntAuthenticate() throws Exception {
+        provider.setSession(userSessionInfo);
+        
+        assertEquals(userSessionInfo, provider.getSession());
+        assertEquals(userSessionInfo, provider.retrieveSession());
+    }
+    
+    @Test
+    public void getSessionReturnsNullIfNotSet() {
+        assertNull(provider.getSession());
+    }
+    
+    @Test
+    public void retrieveSessionWithouSessionSignsInAgain() throws Exception {
+        doReturn(call).when(authenticationApi).signIn(any(SignIn.class));
+        doReturn(response).when(call).execute();
+        doReturn(userSessionInfo).when(response).body();
+        
+        UserSessionInfo retrieved = provider.retrieveSession();
+        assertEquals(userSessionInfo, retrieved);
+        verify(authenticationApi).signIn(any(SignIn.class));
+    }
+    
+    @Test
+    public void reauthenticate() throws Exception {
+        doReturn(call).when(authenticationApi).reauthenticate(any(ReauthenticateRequest.class));
+        doReturn(response).when(call).execute();
+        doReturn(userSessionInfo).when(response).body();
+        
+        userSessionInfo.setReauthToken("reauthToken");
+        provider.setSession(userSessionInfo);
+        
+        provider.reauthenticate();
+        
+        verify(authenticationApi).reauthenticate(reauthenticateRequestCaptor.capture());
+        ReauthenticateRequest reauthRequest = reauthenticateRequestCaptor.getValue();
+        assertEquals("reauthToken", reauthRequest.getReauthToken());
+        assertEquals(signIn.getStudy(), reauthRequest.getStudy());
+        assertEquals(signIn.getEmail(), reauthRequest.getEmail());
+    }
+    
+    @Test
+    public void failureToReauthenticateSignsIn() throws Exception {
```

Comment:
Could you please add a test for when reauth fails and returns a 401, verifying only one reauth attempt is made?

Reply:
Added simulation of getting a 401 back from the server (should switch to sign in on a failed reauth, for now).